src/streamlit_app.py

```diff
@@ -262,6 +262,55 @@ async def draw_messages(
                             status.write(tool_result.content)
                             status.update(state="complete")
 
+            case "custom":
+                try:
+                    task_data = TaskMessageData.model_validate(msg.custom_data)
+                except Exception:
+                    continue
+
+                # If we're rendering new messages, store the message in session state
+                if is_new:
+                    st.session_state.messages.append(msg)
+
+                # If the last message type was not Task, create a new chat message
+                # and container for task messages
+                if last_message_type != "task":
+                    last_message_type = "task"
+                    st.session_state.last_message = st.chat_message(
+                        name="task", avatar=":material/manufacturing:"
+                    )
+                    with st.session_state.last_message:
+                        status = st.status("")
+                    task_messages: dict[str, ChatMessage] = {}
+
+                match task_data.state:
+                    case "new":
+                        status.write(f"Task **{task_data.name}** has :blue[started]. Input:")
+                    case "running":
+                        status.write(f"Task **{task_data.name}** wrote:")
+                    case "complete":
+                        result_text = (
+                            ":green[completed successfully]"
+                            if task_data.result == "success"
+                            else ":red[ended with error]"
+                        )
+                        status.write(f"Task **{task_data.name}** {result_text}. Output:")
+                status.write(task_data.data)
+                status.write("---")
+                task_messages[task_data.run_id] = msg
+                state = "complete"
+                for message in task_messages.values():
```

Comment:
`message` isn't actually referenced here. I was looking through this whole section of code and I'm not sure it works the way you intended - probably can use more cleanup